### PR TITLE
[READY] The Portable Chemical Mixer

### PR DESCRIPTION
**Pull request by LordVollkorn**
## About The Pull Request

Adds the Portable Chemical Mixer to the game, an item that can be worn on the belt slot and dispenses and mixes chemicals. **This item does not produce reagents on its own. Every single reagent has to be supplied to it.** Using a screwdriver, the player can open the device to place in a maximum of 50 beakers, similar to the chemical bag. When closed and in usage, each beaker will appear as a usable button that states the primary reagent of said beaker, along with its remaining units. If a beaker contains more than one reagent, its respective button will dispense all of them in equal amounts. (In Cobbys words: The gunk you put in is the same gunk you get out. -> Your individual beakers dont get filtered.)
The Portable Chemical Mixer is printable at the Medical Techfab and starts out empty.

## Why It's Good For The Game

This item offers an alternative style of gameplay that gives Chemists the incentive to leave their pharmacy and interact with other departments closer, better fulfilling the needs of e.g. the kitchen, botany or the janitor while not having to constantly run back to their dispensers. In the same way, Chemists are now more mobile in finding and receiving reagents (e.g. from Botany). The Portable Chemical Mixer is designed to be future-proof for any possible changes regarding Chemistry.

## Changelog
:cl: Jay Smith & Michael Trauf
add: Portable Chemical Mixer (Requires a vortex anomaly core to activate)
imageadd: Unique sprites for the Portable Chemical Mixer
code: Added Portable Chemical Mixer to the Medical Techfab
/:cl:

### CODE REVIEW

**Comment by Angustmeta** [comment]
Will plastic bottles fit? 

I really miss plastic bottle being a beaker subtype.

**Comment by optimumtact** [comment]
This should require an anomaly core to make.

Pick whatever one suits you best that isn't used

See: https://hackmd.io/@tgstation/r1tzxpwPL

Available types:
https://github.com/tgstation/tgstation/blob/6b424e14a335849038fb762b9b4f580d1f58829b/code/__DEFINES/research/anomalies.dm#L1

**Comment by Ryll-Ryll** [comment]
Don't you think 50 beakers is a bit excessive

**Comment by evilcyborgmrt** [comment]
O man I really like your plans for anomaly cores. According to your design doc, other items that should require cores are phazon, beam rifle and BoH.
Would you really rank this proposed chem belt among those items? It doesn't even do anything you can't already do with a chemistry bag, just streamlines the process.
This thing should obviously be more rare than a chem bag, but requiring an anomaly core is way too much. I'd rather use my core for a BoH so I can carry infinity beakers instead of 50.

**Comment by antropod** [comment]
Anomaly core for everything!
It does synthesize chemicals, just makes mixing easier. What is the problem?

**Comment by antropod** [comment]
> Don't you think 50 beakers is a bit excessive
What amount do you think is acceptable?

**Comment by optimumtact** [comment]
It's not a request it's a precondition for merge

**Comment by antropod** [comment]
What is the reasoning? I don't understand. It's not BoH or phason, it's just a tool for mixing chems, like chem bag with ui.

**Comment by optimumtact** [comment]
It's a portable chem mixer, it's undeniably useful and powerful, I want to be able to precisely control how many of them exist in the game in a round, and the anomaly system allows me to do that.

**Comment by antropod** [comment]
It can not produce chemicals, you still need dispenser to do that.

**Comment by antropod** [comment]
@ExcessiveUseOfCobblestone your thoughts on this? We removed its ability of producing chems from thin air. We made it not purify chemicals.

**Comment by optimumtact** [comment]
I can see that, but it's still a portable mixer, nothing like this has existed before, so I am not going to take any chances.

Lot easier for people to swallow the removal of the anomaly core requirement post merge if we find it's not actually strong/utility than it will be to add it on after the fact.

Easiest route for me is to require it now and review it after a period of use.

**Comment by antropod** [comment]
My problem with anomaly cores is that it makes you dependant on RD.

**Comment by optimumtact** [comment]
technically also cargo, but you should be able to curry favour with both.

**Comment by Drackzo** [comment]
Compromise: it could start with a low beaker count (5 maybe?) and if you want more you need to upgrade it with an anomaly core and maybe better matter bins? I feel locking it completely behind anomaly cores will render it undesirable and you won't get much feedback on whether it's OP or not since no one will use it.

**Comment by optimumtact** [comment]
Listen, I'm open to changing it after it's been in the game and had some time to play out, but I'm aiming to over shoot on initial restriction because in my experience it's a lot easier to buff something after the fact than it is to nerf it.

**Comment by Doosteroni** [comment]
No one:

Oranges: ANOMALY CORE!

**Comment by optimumtact** [comment]
You can expect to see that on a lot of PR's

**Comment by LordVollkorn** [comment]
The Portable Chemical Mixer now requires a vortex anomaly core to function, also reflected in its description

**Comment by OverlyGeneric** [comment]
ah yes, advanced bombing for a glorious death.

**Comment by ATH1909** [comment]
> The Portable Chemical Mixer now requires a vortex anomaly core to function, also reflected in its description

Wouldn't a pyro core be more fitting?

**Comment by ExcessiveUseOfCobblestone** [comment]
> 
> 
> Listen, I'm open to changing it after it's been in the game and had some time to play out, but I'm aiming to over shoot on initial restriction because in my experience it's a lot easier to buff something after the fact than it is to nerf it.

I believe this as well so im fine with core first and followup after we can gauge its strength.

My guess is the only real pro to this compared to say a premade mix would be the plausible deniability.

If I see you have pills/syringes made with lexorin i know you are probably up to no good. I cannot see that if you just have a belt machine and you immediately make the mix afterwards.

**Comment by Krysonism** [comment]
This will not get made until the entire science department has their own set of reactive armour + a phazon for robotics.

File: tgui/packages/tgui/interfaces/PortableChemMixer.js [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+  const beakerContents = recording
+    && Object.keys(data.recordingRecipe)
+      .map(id => ({
+        id,
+        name: toTitleCase(id.replace(/_/, ' ')),
+        volume: data.recordingRecipe[id],
+      }))
+    || data.beakerContents
+    || [];
+  data.chemicals.sort(byTitle);
```
Comment (stylemistake) [inline tgui/packages/tgui/interfaces/PortableChemMixer.js:30]: sortBy ftom common/collections

Comment (LordVollkorn) [reply tgui/packages/tgui/interfaces/PortableChemMixer.js]: Thank you for the advice, I have replaced the old sorting function with sortBy, it works great 👍

**Comment by Doosteroni** [comment]
> This will not get made until the entire science department have their own set of reactive armour + a phazon for robotics.

Yeah what he said, sci ain’t gonna waste their time for a chembag with a ui

**Review by stylemistake — DISMISSED** [review]
Tgui code approved.

**Comment by craigcalef** [comment]
Is this really a "powerful item" that needs an anomaly core to limit economy when its just a short cut to what you can do now with putting beakers in a chem bag and mixing them by hand?

**Review by ExcessiveUseOfCobblestone — CHANGES_REQUESTED** [review]
You have generated new variables/procs/defines without proper documentation for the autodoc to work.

Please click [here ](https://github.com/tgstation/tgstation/blob/master/.github/AUTODOC_GUIDE.md)for a guide on how to properly document these items.

**Comment by LordVollkorn** [comment]
> You have generated new variables/procs/defines without proper documentation for the autodoc to work.

I have added the documentation to my best ability, I hope this meets all the requirements

File: code/game/objects/items/devices/portable_chem_mixer.dm [hunk 2]
```diff
@@ -148,33 +160,33 @@
...
 		is_hallucinating = TRUE
 	for(var/re in dispensable_reagents)
-		var/value = dispensable_reagents[re]
-		var/datum/reagent/temp = GLOB.chemical_reagents_list[re]
+		var/value = dispensable_reagents[re]	///The individual dispensable reagents of the source beakers
+		var/datum/reagent/temp = GLOB.chemical_reagents_list[re]	///The reagents in the source beakers by chemical reagent list
 		if(temp)
-			var/chemname = temp.name
-			var/total_volume = 0
+			var/chemname = temp.name	///The name of the current reagent in question
```
Comment (optimumtact) [inline code/game/objects/items/devices/portable_chem_mixer.dm:181]: To be clear, you only have to document the class level variables, variables defined inside a method are not interesting for documentation.

You should only put comments here if they are going to help the understanding of the process

Comment (LordVollkorn) [reply code/game/objects/items/devices/portable_chem_mixer.dm]: I apologize, I was not familiar with all the terminology at first (and still new to byond coding in general)
This time I only commented the class level variables and the two /proc/'s in the code, and I hope everything is correct now 😊

**Comment by optimumtact** [comment]
I'm going to run a poll sometime this week asking people if they have experienced this yet and what they think.

If people are generally positive we can lift the core requirement

**Comment by LordVollkorn** [comment]
> I'm going to run a poll sometime this week asking people if they have experienced this yet and what they think.
> If people are generally positive we can lift the core requirement

Thank you, I am interested on their feedback as well 👍

**Pull request merged**